pg_replicate/tests/common/sink.rs

```diff
@@ -0,0 +1,169 @@
+use async_trait::async_trait;
+use pg_replicate::conversions::cdc_event::CdcEvent;
+use pg_replicate::conversions::table_row::TableRow;
+use pg_replicate::pipeline::sinks::{BatchSink, InfallibleSinkError};
+use pg_replicate::pipeline::PipelineResumptionState;
+use postgres::schema::{TableId, TableSchema};
+use std::cmp::max;
+use std::collections::{HashMap, HashSet};
+use std::sync::{Arc, Mutex};
+use tokio_postgres::types::PgLsn;
+
+/// A test sink that captures replication events and data for verification.
+///
+/// This sink is designed to be shared across multiple pipelines, simulating
+/// persistent storage while maintaining thread safety through interior mutability.
+#[derive(Debug, Clone)]
+pub struct TestSink {
+    inner: Arc<Mutex<TestSinkInner>>,
+}
+
+/// Internal state of the test sink.
+///
+/// This struct maintains the sink's state including table schemas, rows,
+/// CDC events, and tracking information for copied and truncated tables.
+#[derive(Debug)]
+struct TestSinkInner {
+    // We have a Vec to store all the changes of the schema that we receive over time.
+    tables_schemas: Vec<HashMap<TableId, TableSchema>>,
+    tables_rows: HashMap<TableId, Vec<TableRow>>,
+    events: Vec<Arc<CdcEvent>>,
+    copied_tables: HashSet<TableId>,
+    truncated_tables: HashSet<TableId>,
+    last_lsn: u64,
+}
+
+impl TestSink {
+    /// Creates a new test sink with an empty state.
+    pub fn new() -> Self {
+        Self {
+            inner: Arc::new(Mutex::new(TestSinkInner {
+                tables_schemas: Vec::new(),
+                tables_rows: HashMap::new(),
+                events: Vec::new(),
+                copied_tables: HashSet::new(),
+                truncated_tables: HashSet::new(),
+                last_lsn: 0,
+            })),
+        }
+    }
+
+    /// Updates the last LSN based on received events.
+    ///
+    /// This method ensures that the last LSN is monotonically increasing,
+    /// taking the maximum between the current LSN and the maximum LSN from
+    /// the received events.
+    fn receive_events(&mut self, events: &[CdcEvent]) {
+        let mut max_lsn = 0;
+        for event in events {
+            if let CdcEvent::Commit(commit_body) = event {
+                max_lsn = max(max_lsn, commit_body.commit_lsn());
+            }
+        }
+
+        // We update the last lsn taking the maximum between the maximum of the event stream and
+        // the current lsn, since we assume that lsns are guaranteed to be monotonically increasing,
+        // so if we see a max lsn, we can be sure that all events before that point have been received.
+        let last_lsn = self.inner.lock().unwrap().last_lsn;
+        self.inner.lock().unwrap().last_lsn = max(last_lsn, max_lsn);
```

Comment:
Although this is test code, but can avoid locking twice here.

Reply:
Yup, I did it on purpose to streamline it but yeah, I can squeeze some optimizations out.